ocaml/xapi/storage_smapiv3_migrate.ml

```diff
@@ -12,38 +12,309 @@
+    let request =
+      Http.Request.make
+        ~query:(Http.Url.get_query_params dest_url)
+        ~version:"1.0" ~user_agent:"export_nbd_proxy" Http.Put uri
+    in
+    D.debug "%s making request to dest %s" __FUNCTION__
+      (Http.Url.to_string dest_url) ;
+    let verify_cert = if verify_dest then Stunnel_client.pool () else None in
+    let transport = Xmlrpc_client.transport_of_url ~verify_cert dest_url in
+    with_transport ~stunnel_wait_disconnect:false transport
+      (with_http request (fun (_response, s) ->
+           D.debug "%s starting proxy" __FUNCTION__ ;
+           Unixext.proxy (Unix.dup s) (Unix.dup nbd_client)
+       )
+      ) ;
+    Unix.close proxy_srv
+  with e ->
+    D.debug "%s did not get connection due to %s, closing" __FUNCTION__
+      (Printexc.to_string e) ;
+    Unix.close proxy_srv ;
+    raise e
+
+let mirror_wait ~dbg ~sr ~vdi ~vm ~mirror_id mirror_key =
+  let rec mirror_wait_rec key =
+    let {failed; complete; progress} : Mirror.status =
+      Local.DATA.stat dbg sr vdi vm key
+    in
+    if complete then (
+      Option.fold ~none:()
+        ~some:(fun p -> D.info "%s progress is %f" __FUNCTION__ p)
+        progress ;
+      D.info "%s completed" __FUNCTION__
+    ) else if failed then (
+      Option.iter
+        (fun (snd_state : State.Send_state.t) -> snd_state.failed <- true)
+        (State.find_active_local_mirror mirror_id) ;
+      D.info "%s failed" __FUNCTION__ ;
+      raise
+        (Storage_interface.Storage_error
+           (Migration_mirror_failure "Mirror failed during syncing")
+        )
+    ) else (
+      Option.fold ~none:()
+        ~some:(fun p -> D.info "%s progress is %f" __FUNCTION__ p)
+        progress ;
+      mirror_wait_rec key
+    )
+  in
+
+  match mirror_key with
+  | Storage_interface.Mirror.CopyV1 _ ->
+      ()
+  | Storage_interface.Mirror.MirrorV1 _ ->
+      D.debug "%s waiting for mirroring to be done" __FUNCTION__ ;
+      mirror_wait_rec mirror_key
+
 module MIRROR : SMAPIv2_MIRROR = struct
   type context = unit
 
   let u x = raise Storage_interface.(Storage_error (Errors.Unimplemented x))
 
-  let send_start _ctx = u __FUNCTION__
+  let send_start _ctx ~dbg ~task_id:_ ~dp ~sr ~vdi ~mirror_vm ~mirror_id
+      ~local_vdi:_ ~copy_vm:_ ~live_vm ~url ~remote_mirror ~dest_sr ~verify_dest
+      =
+    let nbd_proxy_path =
+      Printf.sprintf "/var/run/nbdproxy/export/%s" (Vm.string_of mirror_vm)
+    in
+    match remote_mirror with
+    | Mirror.Vhd_mirror _ ->
+        raise
+          (Storage_error
+             (Migration_preparation_failure
+                "Incorrect remote mirror format for SMAPIv3"
+             )
+          )
+    | Mirror.SMAPIv3_mirror {nbd_export; mirror_datapath; mirror_vdi} -> (
+      try
+        let nbd_uri =
+          Uri.make ~scheme:"nbd+unix" ~host:"" ~path:nbd_export
+            ~query:[("socket", [nbd_proxy_path])]
+            ()
+          |> Uri.to_string
+        in
+        let _ : Thread.t =
```

Comment:
This means the current thread has little control over the execution of the split thread. How is it ensured that the thread at the very least times out?

Reply:
Good point, do you have any suggestions... Use some shared variable for communcations?

Reply:
I'm not sure either, we expect the stat to fail at some point while doing the waiting, a timeout could be added to the wait function if there's no progress made over a long period of time, but we've known migration that take 20 hours, so adding a timeout might interrupt long migrations.

The wait loop should also have sort of sleep before calling the next iteration

Reply:
I reckon we should leave this to the `Data.stat` call since it is responsible for detecting anything wrong with the migration. If things go wrong, like the thread exiting abnormally, then `Data.stat` should return a value saying that the mirror has failed, which will result in a `Migration_mirror_failure` being raised